src/Microsoft.VisualStudio.Threading.Shared/AsyncBarrier.cs

```diff
@@ -52,21 +50,33 @@ public AsyncBarrier(int participants)
         /// <returns>An awaitable.</returns>
         public Task SignalAndWait()
         {
-            var tcs = new TaskCompletionSource<bool>();
-            this.waiters.Push(tcs);
-            if (Interlocked.Decrement(ref this.remainingParticipants) == 0)
+            lock (this.waiters)
```

Comment:
Yes, the allocations is what I was trying to avoid regressing. I'd be willing to consider a lock-free approach, but I'd like to see the contention in RPS first. As it is, we'd be trading pessimistic locking for optimistic locking (interlock compare exchange then retry when we lose the race). Both have their costs.
Also, as far as I can find, we don't have *any* users of AsyncBarrier outside of our own Threading tests. So it doesn't seem like a super high priority to optimize. I just wanted to make sure it wasn't buggy so when folks start using it, it is reliable for them.